demos/common/utilities/jws_using_mbedtls.c

```diff
@@ -0,0 +1,634 @@
+/* Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License. */
+
+#include "jws.h"
+
+#include "azure/az_core.h"
+#include "azure/az_iot.h"
+
+#include "azure_iot_result.h"
+#include "azure_iot_json_reader.h"
+#include "azure_iot_adu_client.h"
+
+#include "mbedtls/base64.h"
+#include "mbedtls/rsa.h"
+#include "mbedtls/pk.h"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/cipher.h"
+
+/* For logging */
+#include "demo_config.h"
+
+/**
+ * @brief Convenience macro to return if an operation failed.
+ */
+#define azureiotresultRETURN_IF_FAILED( exp )        \
+    do                                               \
+    {                                                \
+        AzureIoTResult_t const _xAzResult = ( exp ); \
+        if( _xAzResult != eAzureIoTSuccess )         \
+        {                                            \
+            return _xAzResult;                       \
+        }                                            \
+    } while( 0 )
+
+#define jwsRSA3072_SIZE            384
+#define jwsSHA256_SIZE             32
+#define jwsPKCS7_PAYLOAD_OFFSET    19
+
+#define jwsSHA256_JSON_VALUE       "sha256"
+#define jwsSJWK_JSON_VALUE         "sjwk"
+#define jwsKID_JSON_VALUE          "kid"
+#define jwsN_JSON_VALUE            "n"
+#define jwsE_JSON_VALUE            "e"
+#define jwsALG_JSON_VALUE          "alg"
+
+static unsigned char ucJWSHeader[ 1400 ];
```

Comment:
I'm worried that this data is always going to be on the .bss, even when the vast majority of the time it's not actually going to be used.  And it's about 2KB tax here.

I get this complicates the API, but would we be better off having the application to pass in a scratch buffer (and requiring it be some size, also passed in, we could have a "define REQUIRED_SIZE <...>" that they could just use to know right thing ).  

In this case they could decide whether they'd want to allocate all this off the stack (so it'd be temporary) or heap or wherever made most sense for them. 

Vaguely I recall Himanshu did something similar to avoid a malloc() in product code.

Reply:
Another advantage of passing in is that this code as-is isn't thread safe.  I get it's unlikely multiple ADU's would happen at same time, unless... there was an ADU update for a device identity and module going simultaneously.  (Maybe the module represents some peripheral or something that's tied into ADU).  If someone had a scheme like that, it's not inconceivable that on a reconnect then the app would fire ADU off for both module+device identities right away and you could get into this state.